### PR TITLE
Fix workload removal from VIP

### PR DESCRIPTION
**Pull request by howardjohn**
Previously VIP always grew

### CODE REVIEW

File: src/workload.rs [hunk 1]
```diff
@@ -329,6 +329,8 @@ impl WorkloadInformation {
 #[derive(serde::Serialize, Default, Debug)]
 pub struct WorkloadStore {
     workloads: HashMap<IpAddr, Workload>,
+    // workload_to_vip maintains a mapping of workload IP to VIP. This is used only to handle removals.
+    workload_to_vip: HashMap<IpAddr, HashSet<(SocketAddr, u16)>>,
```
Comment (zhlsunshine) [inline src/workload.rs:333]: Shall we consider the dual stack in advance? If yes, we may need to do reasonable design for it.

Comment (hzxuzhonghu) [inline src/workload.rs:333]: I think we can record the vips in `Workload`, then can get ride of this map

Comment (howardjohn) [reply src/workload.rs:333]: We need a way to efficiently lookup a VIP though

Comment (howardjohn) [reply src/workload.rs:333]: What is the issue? IpAddr and SocketAddr can both be IPv4 or IPv6, right?

**Review by stevenctl — APPROVED** [review]

File: src/workload.rs (hunk 1 above)
Comment (zhlsunshine) [reply src/workload.rs:333]: I mean that a workload with IPv4 and IPv6 would have 2 entries in `workloads`, is that what we want?

Comment (zhlsunshine) [reply src/workload.rs:333]:
```
pub struct AdsClient {
    config: Config,
    /// Stores all known workload resources (by name)
    workloads: HashSet<String>,
...
```
As you can see, the workloads are identified by name in `AdsClient`, I think there may be an issue.

Comment (howardjohn) [reply src/workload.rs:333]: Yes, we want 2 workloads in that case IMO. `name` is IP

Comment (howardjohn) [reply src/workload.rs:333]: Well its a bit confusing. ResourceName == IP, but `name` is the pod name

Comment (zhlsunshine) [reply src/workload.rs:333]: yes, shall I change it as?
```
workloads: HashSet<IpAddr>,
```

Comment (zhlsunshine) [reply src/workload.rs:333]: Sorry, @howardjohn, I should do this change:
```
pub struct ResourceKey {
    pub resource_name: String,
    pub type_url: String,
}
```

Comment (howardjohn) [reply src/workload.rs:333]: No it's good as is

Comment (zhlsunshine) [reply src/workload.rs:333]: and 
```
pub struct XdsResource<T: prost::Message> {
    pub resource_name: String,
    pub resource: T,
}
```

Comment (howardjohn) [reply src/workload.rs:333]: I'm not sure I understand what you mean

Comment (zhlsunshine) [reply src/workload.rs:333]: > Well its a bit confusing. ResourceName == IP, but `name` is the pod name

I understand that everything is okay for now except there are some field name confusing in `ResourceKey` definition. 
If okay, no anything need to be changed here.

**Pull request merged**